src/middlewared/middlewared/etc_files/local/netdata/alarms.py

```diff
@@ -1,26 +1,38 @@
+import os
 import re
 
 
 async def render(service, middleware):
     def update_alarms(alarms):
-        for key, alarm in alarms.items():
-
-            # These are valid alarms and their respective config files exist as well
-            if alarm.get('path'):
-                with open(alarm['path'], 'r+') as f:
+        for path, alarm_dict in alarms.items():
+            if os.path.exists(path):
+                with open(path, 'r') as f:
                     content = f.read()
-                    f.seek(0)
-                    f.write(
-                        re.sub(
-                            fr'(alarm: {key}[\s\S]*?to: )(.*)',
-                            fr'\1{"silent" if not alarm["enabled"] else "sysadmin"}',
-                            content
-                        )
+
+                alarm = None
+                for key, alarm in alarm_dict.items():
+                    content = re.sub(
+                        fr'(alarm: {key}[\s\S]*?to: )(.*)',
+                        fr'\1{"silent" if not alarm["enabled"] else "sysadmin"}',
+                        content
                     )
-                    f.truncate()
+                # Better safe then sorry
+                if alarm:
+                    with open(alarm['write_path'], 'w') as f:
+                        f.write(content)
+
             else:
-                middleware.logger.debug(f'Could not find config file for {key} alarm')
+                middleware.logger.debug(
+                    f'Could not find config file {path} for {",".join(list(alarm_dict))} alarm(s)'
+                )
+
+    listed_alarms = await middleware.call('netdata.list_alarms')
+    alarms = {}
+    # Let's not unnecessarily open a single file again and again for reading/writing
+    for key, alarm in listed_alarms.items():
+        if not alarms.get(alarm.get('read_path')):
+            alarms[alarm.get('read_path')] = {}
```

Comment:
Seems like these couple lines could removed if you used a defaultdict.

from collections import defaultdict

```
alarms = defaultdict(dict)
for key, alarm in listed_alarms.items():
    alarms[alarm.get('read_path')][key] = alarm
```